### ISSUE TITLE
CUB `device_reduce` and `size_t`

### ISSUE THREAD

**Issue by emcastillo** [cupy/cupy#3309]
cc @leofang @jakirkham 

Continuing the discussion in #3117

**Comment by leofang** [comment]
@kmaehashi This is not a bug in CuPy, but at most a "defect" in CUB. I think it is best to bring this  issue to upstream (now it's [thrust/cub](https://github.com/thrust/cub)) and ask the current maintainers to address it.

**Comment by jakirkham** [comment]
Yes and no.

Basically what they have said is they don't want to use a 64-bit type (sticking with a 32-bit type) due to register pressure. So this is something that we would have to workaround or override. This is is the "no" part 😉

That said, they are choosing to use a signed type where an unsigned type seems more appropriate. That sounds like a bug, but IDK what their plans are for addressing that.

**Comment by leofang** [comment]
> Basically what they have said is they don't want to use a 64-bit type (sticking with a 32-bit type) due to register pressure. So this is something that we would have to workaround or override. This is is the "no" part 😉

1. From what an outsider can tell, I think the "they" has changed, so the new "they" may have different opinions 😉
2. If they do what I suggested (templated type), it wouldn't affect the normal usage or performance unless downstream consumers (i.e. us) explicitly overwrites it.  

> That said, they are choosing to use a signed type where an unsigned type seems more appropriate. That sounds like a bug, but IDK what their plans are for addressing that. Though this may be something we can override as well.

This will also be fixable with templated type.

**Comment by jakirkham** [comment]
Sorry for the lack of clarity. They means CUB developers.

**Comment by leofang** [comment]
Yeah I know, and I meant it seems Thrust folks have taken over the development & maintenance work from Duane?

**Comment by jakirkham** [comment]
I also discussed this with Bryce offline.

**Comment by leofang** [comment]
Cool. Keep us posted. If needed I can also raise a counterpart of NVlabs/cub#129 to thrust/cub.

**Comment by jakirkham** [comment]
The response was more-or-less the same. So I don't have more to add.

There doesn't appear to be an issue tracker for thrust/cub though.

**Comment by leofang** [comment]
Ah, they turned it off...Alright. What's their opinion toward a templated type for item size?

**Comment by leofang** [comment]
I hope I am pinging the right Bryce 😂 @brycelelbach, any comment for my suggestion to fix the issue on the item size type that John raised?

**Comment by brycelelbach** [comment]
thrust/cub is just a temporary repo until we get a new GitHub org for both. We didn't want to turn issues on there. I hope to have this sorted out soon.

**Comment by brycelelbach** [comment]
@griwes recently did a bunch of work to fix 32bit indexing issues in Thrust/CUB, I'll ask him to take a look at this.

**Comment by leofang** [comment]
Thanks, Bryce! Please keep us posted. 🙏

**Comment by griwes** [comment]
Hi folks,

I appreciate the suggestion mentioned above of "just" adding a template parameter and deducing a possibly larger size type, but not doing that was a conscious decision made by Duane (the original author of CUB), and I fully agree with it. I used to not be sure about it, but I've made a change that made certain Thrust algorithms unconditionally made a call into CUB with a 64-bit size type (I'll explain how in a moment), and benchmarked it against the "default" implementation... aaaaand it was about 10% slower. This may be a source-compatible change, but it's not ABI compatible, and it sure as day isn't performance-compatible ;)

Take the following preexisting (pseudo)code:

```cpp
std::size_t size = ...;
char * buf = obtain_buffer(size);
cub::DeviceReduce::Reduce(..., buf, ..., size, ...);
```

This code used to use 32-bit signed integers as the indices; with the change you're suggesting, it now uses 64-bit unsigned integers. This is going to be a performance hit due to register pressure, _and_ potentially a performance hit because the compiler is no longer allowed to assume that wraparound does not happen. Oops.

Now, was this code correct? Who knows. The problem is that a lot of code just indiscriminantly uses `std::size_t` for sizes, even if their problem sizes never approach the range of values representable by those types. I don't think that problem sizes upwards of 2^31-1 will be the majority of uses for CUB, so I don't believe that making those cases more convenient than the ones that do fit in an `int` is reasonable.

Now, CUB has two different layers of dispatching. Note how [the full implementation of `DeviceReduce::Reduce`](https://github.com/thrust/cub/blob/master/cub/device/device_reduce.cuh#L162) is actually... just a call to `DispatchReduce<InputIteratorT, OutputIteratorT, OffsetT, ReductionOpT>::Dispatch`. I don't think this layer is documented; we should probably think about doing that in the (near?) future. But it's the key to achieving what you want, and to what I did to make sure the Thrust uses of CUB algorithms don't exhibit weird bugs at 2^31-1 and up.

To dispatch across the different implementations based on the size at runtime, I used [this simple dispatching macro](https://github.com/thrust/thrust/blob/master/thrust/system/cuda/detail/dispatch.h#L58-L78) (hopefully the comment quality in that file is good enough for you to grok the idea behind them; I tried to leave a trace that details what is going on). The use of it for reduce specifically is [here](https://github.com/thrust/thrust/blob/master/thrust/system/cuda/detail/reduce.h#L945-L954) (where `Size` in this context is actually deduced from the call site). On the path on which we know the size fits into `int32_t`, we just call the regular form of the CUB `DeviceReduce` entrypoint. On the path where we detected a size larger than the capacity of `int32_t`, we invoke the `DispatchReduce` entrypoint, forcing the size type to be what was deduced.

This may or may not be what you want to do; if you don't care about the absolutely highest performance possible, and do not want to compile all the CUB reduce kernels twice (for two different size types, which is the price we pay for having both correctness for large problem sizes **and** performance that is consistent with prior releases), you may just want to call into `DispatchReduce<..., ..., std::size_t, ...>::Dispatch`.

This is very similar for the other CUB algorithms.

Hopefully this helps ;)

**Comment by jakirkham** [comment]
Thanks Michał! So one question that comes up is why `int32_t` is used as opposed to `uint32_t`? Seems we would get 2x more values without leaving 32-bits. Guessing there is just something I'm not understanding here. Any thoughts you have would be appreciated 🙂

**Comment by leofang** [comment]
Hi @griwes, thanks a lot for your investigation and reply! In addition to John's nice question [stemmed from our previous discussion](https://github.com/NVlabs/cub/issues/129#issuecomment-591055945), I have a few more questions:

> I used to not be sure about it, but I've made a change that made certain Thrust algorithms unconditionally made a call into CUB with a 64-bit size type (I'll explain how in a moment), and benchmarked it against the "default" implementation... aaaaand it was about 10% slower. 

Yes, we learned this fact from Duane's reply in https://github.com/NVlabs/cub/issues/129#issuecomment-393296009, and I myself also observed this register pressure issue in my CUB-related work.

[What I suggested](https://github.com/cupy/cupy/issues/3309#issuecomment-620654778) is to add a new template parameter to size type, but default it to `int` so as to ensure backward compatibility (in terms of both source and performance). So, for what you mentioned next

> This may be a source-compatible change, but it's not ABI compatible, and it sure as day isn't performance-compatible ;)

I have trouble to understand it. Could you elaborate everything here (source/ABI/performace)? 

> Now, was this code correct? Who knows. The problem is that a lot of code just indiscriminantly uses `std::size_t` for sizes, even if their problem sizes never approach the range of values representable by those types. I don't think that problem sizes upwards of 2^31-1 will be the majority of uses for CUB, so I don't believe that making those cases more convenient than the ones that do fit in an `int` is reasonable.

To be clear, I am fully aware that using `size_t` is gonna hurt performance. The problem for library providers like CuPy is that there will always be users (aka John 😂) asking for this capability. Some of NVIDIA GPUs also have large enough memory to actually back arrays of size > 2^32 and small types (ex: half/int/float), so it's not physically impossible. Additional constraint for CuPy is it must be NumPy-complaint whenever possible, and NumPy does allow reductions over 2^31-1 elements. So, we need the freedom (not that we like it or would always use it) to sacrifice a little to some performance in exchange for better compliance and/or extended support, because *there are use cases*.

> Now, CUB has two different layers of dispatching. Note how [the full implementation of `DeviceReduce::Reduce`](https://github.com/thrust/cub/blob/master/cub/device/device_reduce.cuh#L162) is actually... just a call to `DispatchReduce<InputIteratorT, OutputIteratorT, OffsetT, ReductionOpT>::Dispatch`. I don't think this layer is documented; we should probably think about doing that in the (near?) future. 

Yes, declaring this layer as a part of stable, public API would make sense if you want advanced users to start from here; working on top of undocumented 3rd party API is not acceptable in production code, after all. 

**Actually, what you suggested is a variant of my suggestion: the interface is already there, all we need to is to make it public and defaulted to `typename OffsetT=int`.** Of course, it's even better if you expose `OffsetT` one layer up to `DeviceReduce` and `DeviceSegmentedReduce` (and apply the default).

> To dispatch across the different implementations based on the size at runtime, I used [this simple dispatching macro](https://github.com/thrust/thrust/blob/master/thrust/system/cuda/detail/dispatch.h#L58-L78) (hopefully the comment quality in that file is good enough for you to grok the idea behind them; I tried to leave a trace that details what is going on). The use of it for reduce specifically is [here](https://github.com/thrust/thrust/blob/master/thrust/system/cuda/detail/reduce.h#L945-L954) (where `Size` in this context is actually deduced from the call site). On the path on which we know the size fits into `int32_t`, we just call the regular form of the CUB `DeviceReduce` entrypoint. On the path where we detected a size larger than the capacity of `int32_t`, we invoke the `DispatchReduce` entrypoint, forcing the size type to be what was deduced.

Thanks for the pointers, Michał! It's very useful for me. (I've been fairly familiar with CUB internals but it's less easy for me to dive into Thrust...)

However, for CuPy we have a simpler solution. All of our arrays live in Python and we know their sizes prior to invoking CUDA kernels, so the dispatcher could just live at the Cython level (or C++ as we currently do, it doesn't really matter). But in order to reduce performance hit for smaller arrays, we'll need the path splitting like yours indeed (including the double compilation you mentioned below), and it's actually what's already done in CuPy's internal (not involving CUB).

> This may or may not be what you want to do; if you don't care about the absolutely highest performance possible, and do not want to compile all the CUB reduce kernels twice (for two different size types, which is the price we pay for having both correctness for large problem sizes **and** performance that is consistent with prior releases), you may just want to call into `DispatchReduce<..., ..., std::size_t, ...>::Dispatch`.

I think I replied to this part earlier. I can try to make it clearer if needed.

Thanks again 🙂

**Comment by leofang** [comment]
@griwes @brycelelbach Would be great if you folks could follow up this thread 🙂

@jakirkham @emcastillo The oversize reduction issue could be resolved by #3244.

**Comment by leofang** [comment]
**Update:** in #3562 we will disable large size (>2^31) CUB device reduction routines (those from `cupy.cuda.cub`), as they currently don't work anyway. If `CUPY_ACCELERATORS="cub"` is set, then it will fall back from those routines to the new block reduction implementation #3244, otherwise to CuPy's original reduction implementation. This will be a temporary workaround until CUB makes its internal interface that allows a `size_t` specialization (https://github.com/cupy/cupy/issues/3309#issuecomment-626073805) documented.

**Comment by griwes** [comment]
@leofang I'm unsure why you are blocked by the interface not being documented. It's a part of the public API and it's used by Thrust directly, so it's not going to go away or change.

Sorry I missed your question about suggestions earlier, I could swear I replied some time ago... No, we cannot bubble the template parameter up to the primary interface, because it'd then either:
1. Always be deduced to size_t for most folks because that's what the type they actually pass in is (and template argument deduction is ruthless, default parameters don't matter if they get deduced, or
2. Make that parameter the first template parameter for the primary interface, which would be a source code break if anyone tries to pass template arguments explicitly right now.

We _could_ try to add some _new_ overloads that require you to pass the offset type explicitly, though we'd need to be rather careful to not break any overload sets in the process. Please file an issue on github.com/thrust/cub if you'd like that, but I'd recommend that you proceed with just using the Dispatch layer directly.

**Comment by leofang** [comment]
Hi @griwes Thanks for replying! 

To the public API thing: I personally would really love to take your words and start working on it 🙂 However, we need a stronger commitment from the Thrust/CUB team, in that it needs to be documented on http://nvlabs.github.io/cub/modules.html, which currently is not the case. 

I understand that it being public is the status quo and that Thrust already uses it internally, but still I would like to see a stronger guarantee to downstream libraries, as Thrust/CUB are unified and free to change each other as needed. 

Now, to the promotion to a template parameter, I didn't know this when I replied: 
> template argument deduction is ruthless, default parameters don't matter if they get deduced

I've been bullied by SFINAE lately, so I believe you're right...😂

Another action I'd like to see taken on the CUB side for strengthening the above guarantee is that oversize arrays indexed by `size_t` need to be tested more extensively. A quick grep does not show me any test case for this capability. Given my recent experience with failing to compile the CUB histogram support for `size_t` (which I fix in thrust/cub#38), I incline to think CUB does not treat this issue seriously (which was actually understandable given its design trajectory).

Finally, please poke @brycelelbach for me when you get a chance 😅 and ask him to open the issue tracker on https://github.com/thrust/cub, since it's currently disabled. My impression is that the issue tracker on the old repo NVlabs/CUB is almost abandoned and not monitored...I have a few other CUB-related things that I would like to discuss, but I need to know where to go. Don't wanna keep pinging you folks...

Thanks!

**Comment by leofang** [comment]
Thrust and CUB are addressing this issue on their end: https://github.com/NVIDIA/thrust/issues/1271, https://github.com/NVIDIA/cub/issues/212.
